File: lib/noaa.ts
/**
 * NOAA Space Weather K-Index API Integration
 * Fetches real-time geomagnetic activity data using standard fetch (Node.js runtime)
 * Now with Redis caching for improved performance
 *
 * K-Index Scale (0-9):
 * 0-4: Quiet to unsettled
 * 5: Minor storm
 * 6: Moderate storm
 * 7: Strong storm
 * 8: Severe storm
 * 9: Extreme storm
 */

import { getOrCompute, CacheKeys, CacheTTL } from './redis';

export interface KIndexReading {
  timeTag: string; // ISO 8601 timestamp
  kpIndex: number; // Planetary K-Index (0-9)
  observedTime: Date;
}

export interface NOAASpaceWeatherData {
  latest: KIndexReading;
  readings: KIndexReading[];
  averageKp: number;
  maxKp: number;
  status: 'quiet' | 'unsettled' | 'storm';
  isSimulated?: boolean; // True when NOAA API is unavailable and mock data is used
}

/**
 * Fetch current K-Index from NOAA SWPC JSON API
 * Uses the 3-day K-Index forecast/observation endpoint
 * Results are cached in Redis for 5 minutes
 */
export async function fetchKIndex(): Promise<NOAASpaceWeatherData> {
  return getOrCompute(
    CacheKeys.NOAA_KINDEX,
    fetchKIndexFromAPI,
    CacheTTL.NOAA_DATA
  );
}

/**
 * Direct API fetch (used by caching layer)
 */
async function fetchKIndexFromAPI(): Promise<NOAASpaceWeatherData> {
  try {
    // NOAA SWPC JSON API endpoint for K-Index data
    const response = await fetch(
      'https://services.swpc.noaa.gov/products/noaa-planetary-k-index.json',
      {
        cache: 'no-store', // Ensure fresh data for real-time monitoring
        headers: {
          'Accept': 'application/json',
        },
      }
    );

    if (!response.ok) {
      throw new Error(`NOAA API error: ${response.status} ${response.statusText}`);
    }

    const data: unknown = await response.json();
    
    // Data format: [[header], [timestamp, kp, kp_index], ...]
    // Skip the header row
    if (!Array.isArray(data)) {
      throw new Error('Invalid response format from NOAA API');
    }
    
    const readings: KIndexReading[] = data
      .slice(1) // Skip header
      .filter((row: unknown): row is [string, number, string?] => {
        return Array.isArray(row) && row.length >= 2
          && typeof row[0] === 'string'
          && row[1] !== null && row[1] !== undefined;
      })
      .map((row: [string, number, string?]) => ({
        timeTag: row[0],
        kpIndex: parseFloat(String(row[1])),
        observedTime: new Date(row[0]),
      }))
      .sort((a: KIndexReading, b: KIndexReading) => 
        b.observedTime.getTime() - a.observedTime.getTime()
      );

    if (readings.length === 0) {
      throw new Error('No valid K-Index readings available from NOAA');
    }

    const latest = readings[0];
    
    // Calculate statistics
    const validKpValues = readings.map(r => r.kpIndex).filter(v => !isNaN(v));
    const averageKp = validKpValues.reduce((sum, val) => sum + val, 0) / validKpValues.length;
    const maxKp = Math.max(...validKpValues);
    
    // Determine geomagnetic storm status
    const status = getStormStatus(latest.kpIndex);

    return {
      latest,
      readings: readings.slice(0, 24), // Last 24 readings (8 hours)
      averageKp: Math.round(averageKp * 100) / 100,
      maxKp,
      status,
    };
  } catch (error) {
    console.error('Error fetching K-Index from NOAA:', error);
    
    // Return mock data for development/offline mode
    return getMockKIndexData();
  }
}

/**
 * Determine storm status from K-Index value
 */
function getStormStatus(kp: number): 'quiet' | 'unsettled' | 'storm' {
  if (kp >= 5) return 'storm';
  if (kp >= 4) return 'unsettled';
  return 'quiet';
}

/**
 * Mock data for development and error fallback
 */
function getMockKIndexData(): NOAASpaceWeatherData {
  const now = new Date();
  const mockReadings: KIndexReading[] = [];
  
  // Generate 24 mock readings (past 8 hours, 3-hour intervals)
  for (let i = 0; i < 24; i++) {
    const timeOffset = (23 - i) * 15 * 60 * 1000; // 15 minutes apart
    const timestamp = new Date(now.getTime() - timeOffset);
    
    // Simulate varying K-Index with some randomness
    const baseKp = 3 + Math.sin(i / 4) * 2;
    const kpIndex = Math.max(0, Math.min(9, baseKp + (Math.random() - 0.5)));
    
    mockReadings.push({
      timeTag: timestamp.toISOString(),
      kpIndex: Math.round(kpIndex * 10) / 10,
      observedTime: timestamp,
    });
  }
  
  const latest = mockReadings[mockReadings.length - 1];
  const avgKp = mockReadings.reduce((sum, r) => sum + r.kpIndex, 0) / mockReadings.length;
  const maxKp = Math.max(...mockReadings.map(r => r.kpIndex));
  
  return {
    latest,
    readings: mockReadings,
    averageKp: Math.round(avgKp * 100) / 100,
    maxKp,
    status: getStormStatus(latest.kpIndex),
    isSimulated: true,
  };
}

/**
 * Get human-readable K-Index description
 */
export function getKIndexDescription(kp: number): string {
  if (kp >= 9) return 'Extreme Geomagnetic Storm (G5)';
  if (kp >= 8) return 'Severe Geomagnetic Storm (G4)';
  if (kp >= 7) return 'Strong Geomagnetic Storm (G3)';
  if (kp >= 6) return 'Moderate Geomagnetic Storm (G2)';
  if (kp >= 5) return 'Minor Geomagnetic Storm (G1)';
  if (kp >= 4) return 'Unsettled Conditions';
  return 'Quiet Conditions';
}

/**
 * Get color code for K-Index visualization
 */
export function getKIndexColor(kp: number): string {
  if (kp >= 8) return '#dc2626'; // red-600
  if (kp >= 7) return '#ea580c'; // orange-600
  if (kp >= 6) return '#f59e0b'; // amber-500
  if (kp >= 5) return '#eab308'; // yellow-500
  if (kp >= 4) return '#84cc16'; // lime-500
  return '#10b981'; // emerald-500
}

File: backend/services/noaa.py
"""
NOAA Space Weather K-Index API Integration
Fetches real-time geomagnetic activity data with Redis caching

K-Index Scale (0-9):
0-4: Quiet to unsettled
5: Minor storm
6: Moderate storm
7: Strong storm
8: Severe storm
9: Extreme storm

All response models include both snake_case and camelCase field aliases.
"""

import math
import random
import logging
from datetime import datetime
from typing import List, Literal, Optional
import httpx
from pydantic import BaseModel, Field, ConfigDict
from pydantic.alias_generators import to_camel

from services.redis_cache import get_or_compute, CacheKeys, CacheTTL

logger = logging.getLogger(__name__)


class DualCaseModel(BaseModel):
    """Base model providing both snake_case and camelCase field names"""
    model_config = ConfigDict(
        populate_by_name=True,
        alias_generator=to_camel,
    )

    def model_dump_dual(self, **kwargs) -> dict:
        """Dump model with both snake_case and camelCase fields"""
        snake_dict = self.model_dump(by_alias=False, **kwargs)
        result = {}
        for key, value in snake_dict.items():
            result[key] = value
            # Add camelCase alias
            camel_key = ''.join(
                word.capitalize() if i > 0 else word
                for i, word in enumerate(key.split('_'))
            )
            if camel_key != key:
                result[camel_key] = value
        return result


class KIndexReading(DualCaseModel):
    """Single K-Index reading with dual-case field names"""
    time_tag: str = Field(description="Timestamp from NOAA")
    kp_index: float = Field(description="K-Index value (0-9)")
    observed_time: datetime = Field(description="Parsed observation time")


class NOAASpaceWeatherData(DualCaseModel):
    """Complete NOAA space weather response with dual-case field names"""
    latest: KIndexReading = Field(description="Most recent K-Index reading")
    readings: List[KIndexReading] = Field(description="Array of recent readings")
    average_kp: float = Field(description="Average K-Index value")
    max_kp: float = Field(description="Maximum K-Index value")
    status: Literal["quiet", "unsettled", "storm"] = Field(description="Storm status")
    is_simulated: bool = Field(default=False, description="True if using mock data")


def get_storm_status(kp: float) -> Literal["quiet", "unsettled", "storm"]:
    """Determine storm status from K-Index value"""
    if kp >= 5:
        return "storm"
    if kp >= 4:
        return "unsettled"
    return "quiet"


def get_kindex_description(kp: float) -> str:
    """Get human-readable K-Index description"""
    if kp >= 9:
        return "Extreme Geomagnetic Storm (G5)"
    if kp >= 8:
        return "Severe Geomagnetic Storm (G4)"
    if kp >= 7:
        return "Strong Geomagnetic Storm (G3)"
    if kp >= 6:
        return "Moderate Geomagnetic Storm (G2)"
    if kp >= 5:
        return "Minor Geomagnetic Storm (G1)"
    if kp >= 4:
        return "Unsettled Conditions"
    return "Quiet Conditions"


def get_kindex_color(kp: float) -> str:
    """Get color code for K-Index visualization"""
    if kp >= 8:
        return "#dc2626"  # red-600
    if kp >= 7:
        return "#ea580c"  # orange-600
    if kp >= 6:
        return "#f59e0b"  # amber-500
    if kp >= 5:
        return "#eab308"  # yellow-500
    if kp >= 4:
        return "#84cc16"  # lime-500
    return "#10b981"  # emerald-500


def get_mock_kindex_data() -> dict:
    """Generate mock data for development and error fallback"""
    now = datetime.utcnow()
    mock_readings = []

    for i in range(24):
        time_offset_ms = (23 - i) * 15 * 60 * 1000
        timestamp = datetime.fromtimestamp(now.timestamp() - time_offset_ms / 1000)

        # Simulate varying K-Index with some randomness
        base_kp = 3 + math.sin(i / 4) * 2
        kp_index = max(0, min(9, base_kp + (random.random() - 0.5)))

        mock_readings.append({
            "time_tag": timestamp.isoformat(),
            "kp_index": round(kp_index, 1),
            "observed_time": timestamp.isoformat()
        })

    latest = mock_readings[-1]
    avg_kp = sum(r["kp_index"] for r in mock_readings) / len(mock_readings)
    max_kp = max(r["kp_index"] for r in mock_readings)

    return {
        "latest": latest,
        "readings": mock_readings,
        "average_kp": round(avg_kp, 2),
        "max_kp": max_kp,
        "status": get_storm_status(latest["kp_index"]),
        "is_simulated": True
    }


async def fetch_kindex_from_api() -> dict:
    """Direct API fetch from NOAA"""
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                "https://services.swpc.noaa.gov/products/noaa-planetary-k-index.json",
                headers={"Accept": "application/json"},
                timeout=10.0
            )
            response.raise_for_status()
            data = response.json()

        if not isinstance(data, list):
            raise ValueError("Invalid response format from NOAA API")

        # Skip header row and parse readings
        readings = []
        for row in data[1:]:
            if len(row) >= 2 and row[1] is not None:
                try:
                    kp_value = float(row[1])
                    observed_time = datetime.fromisoformat(row[0].replace("Z", "+00:00").replace(" ", "T"))
                    readings.append({
                        "time_tag": row[0],
                        "kp_index": kp_value,
                        "observed_time": observed_time.isoformat()
                    })
                except (ValueError, IndexError):
                    continue

        if not readings:
            raise ValueError("No valid K-Index readings available from NOAA")

        # Sort by time descending
        readings.sort(key=lambda x: x["observed_time"], reverse=True)

        latest = readings[0]
        valid_kp_values = [r["kp_index"] for r in readings if not math.isnan(r["kp_index"])]
        if not valid_kp_values:
            raise ValueError("All K-Index readings are NaN")
        average_kp = sum(valid_kp_values) / len(valid_kp_values)
        max_kp = max(valid_kp_values)

        return {
            "latest": latest,
            "readings": readings[:24],  # Last 24 readings
            "average_kp": round(average_kp, 2),
            "max_kp": max_kp,
            "status": get_storm_status(latest["kp_index"]),
            "is_simulated": False
        }

    except Exception as e:
        logger.warning(f"Error fetching K-Index from NOAA: {e}")
        return get_mock_kindex_data()


async def fetch_kindex() -> dict:
    """Fetch K-Index with Redis caching"""
    return await get_or_compute(
        CacheKeys.NOAA_KINDEX,
        fetch_kindex_from_api,
        CacheTTL.NOAA_DATA
    )
